source/classes.tex

```diff
@@ -6597,7 +6597,7 @@
 }
 
 constexpr A a;          // well-formed, \tcode{a.p} points to \tcode{a}
-constexpr A b = g();    // error, \tcode{b.p} would point to an automatic object
+constexpr A b = g();    // error, \tcode{b.p} would point to a storage whose duration has ended
```

Comment:
I agree with the rationale in general, but where in [expr.const] do we say this is not a constant expression? Or is the normative text elsewhere? (We're not applying an lvalue-to-rvalue conversion at this point, I believe.)

Reply:
> [expr.const]/(10.2) - if the value is of pointer type, it contains the address of an object with static storage duration, the address past the end of such an object ([expr.add]), the address of a non-immediate function, or a null pointer value,

> [expr.const]/(10.4) - if the value is an object of class or array type, each subobject satisfies these constraints for the value.

As `g()` is a prvalue of class type, its subobject (`g().p`) must satifiy (10.1) - (10.3) to make `g()` a constant expression. However, as NRVO is forbidden, `g().p` points to the local variable with automatic storage duration and fails to satisfy (10.2).

Reply:
Ok, thanks. But the criterion is not "storage has ended", but "static storage duration", it seems.

Suggestion, with a more colloquial phrasing: "error: \tcode{b.p} would be dangling\iref{expr.const}"

Reply:
OK. I've changed the phrasing.